AsyncDisplayKit/Layout/ASLayoutSpec.mm

```diff
@@ -113,28 +112,35 @@ - (void)setParent:(id<ASLayoutable>)parent
   }
 }
 
-- (void)setChild:(id<ASLayoutable>)child;
+- (void)setChild:(id<ASLayoutable>)child
 {
   ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
-  
+  ASDisplayNodeAssertNotNil(child, @"Child cannot be nil");
+
   id<ASLayoutable> finalLayoutable = [self layoutableToAddFromLayoutable:child];
-  _child = finalLayoutable;
-  [self propagateUpLayoutable:finalLayoutable];
+  if (finalLayoutable) {
+    _children = @[finalLayoutable];
+    [self propagateUpLayoutable:finalLayoutable];
+  }
 }
 
 - (void)setChild:(id<ASLayoutable>)child forIdentifier:(NSString *)identifier
 {
   ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
+  ASDisplayNodeAssertNotNil(child, @"Child cannot be nil");
   
   id<ASLayoutable> finalLayoutable = [self layoutableToAddFromLayoutable:child];
   self.childrenWithIdentifier[identifier] = finalLayoutable;
+  if (finalLayoutable) {
+    self.children = [self.children arrayByAddingObject:finalLayoutable];
```

Comment:
This isn't great. Now that `ASLayoutSpec` conforms to `ASEnvironment`, `children` is expected to return all of the subspecs. The dictionary is around because subclasses of `ASLayoutSpec` like `ASBackgroundLayoutSpec`. Instead of the convention of `_children[0]` is foreground and `_children[1]` is background, I created a dictionary to save away the children based on static strings defined in the subclass. 

Now that I'm adding all subspecs to `_children` anyway, the dictionary is around for little more than a look up. So, to answer your question, adding to the children array here is correct. BUT you should hound me to put up another PR cleaning this up so that we don't need the dictionary.